cg/meta/workflow/microsalt.py

```diff
@@ -306,3 +344,96 @@ def get_case_id_from_case(self, unique_id: str) -> Tuple[str, None]:
             raise click.Abort
         case_id = case_obj.internal_id
         return case_id, None
+
+    def microsalt_qc(self, case_id: str, run_dir_path: Path, lims_project: str) -> bool:
+        """Check if Microsalt case passes QC check."""
+        samples: List[Sample] = self.get_samples(case_id=case_id)
+        failed_samples: Dict = {}
+        qc_file: Dict = read_json(file_path=Path(run_dir_path, f"{lims_project}.json"))
+
+        for sample in samples:
+            if not sample.sequenced_at:
+                continue
+            self.qc_sample_check(sample=sample, failed_samples=failed_samples, qc_file=qc_file)
+
+        return self.qc_case_check(
+            case_id=case_id,
+            failed_samples=failed_samples,
+            samples=samples,
+            run_dir_path=run_dir_path,
+        )
+
+    def qc_case_check(
+        self, case_id: str, samples: List[Sample], failed_samples: Dict, run_dir_path: Path
+    ) -> bool:
+        """Perform the final QC check for a microbial case based on failed samples."""
+        qc_pass: bool = True
+
+        for sample in failed_samples:
+            # Check if negative control failed
+            if sample.control == "negative":
+                qc_pass = False
+            # Check if MWR sample failed
+            if sample.application_version.application.tag == MicrosaltAppTags.MWRNXTR003:
+                qc_pass = False
+
+        # Check if more than 10% of MWX samples failed
+        if len(failed_samples) / len(samples) > MicrosaltQC.QC_PERCENT_THRESHOLD_MWX:
+            qc_pass = False
+
+        if not qc_pass:
+            LOG.warning(
+                f"Case {case_id} failed QC, see {run_dir_path}/QC_done.txt for more information."
+            )
+        else:
+            LOG.info(f"Case {case_id} passed QC.")
+
+        self.create_qc_done_file(
+            run_dir_path=run_dir_path,
+            failed_samples=failed_samples,
+        )
+        return qc_pass
+
+    def create_qc_done_file(self, run_dir_path: Path, failed_samples: Dict) -> None:
+        """Creates a QC_done when a QC check is performed."""
+        with open(os.path.join(run_dir_path, "QC_done.txt"), "w") as file:
+            for sample_dict in failed_samples.items():
+                file.write(f"{sample_dict[0].internal_id}: {json.dumps(sample_dict[1])} \n")
+
+    def qc_sample_check(self, sample: Sample, failed_samples: Dict, qc_file: Dict) -> None:
```

Comment:
Can you try to only pass the necessary information to the function, to decouple it more from the context? Here I think one can do:
```suggestion
    def qc_sample_check(self, sample: Sample, sample_qc: Dict) -> Dict:
```
and then where it is called:
```shell
failed_samples[sample.internal_id]=self.qc_sample_check(sample=sample, sample_qc=case_qc[sample.internal_id])
```

Reply:
very nice comment!